File: python/sort/sort_test.py
import unittest
from bubble_sort import bubble_sort_in_place


class TestBubbleSort(unittest.TestCase):
    def test_sort_empty_list(self):
        # Test sorting an empty list
        elements = []
        bubble_sort_in_place(elements)
        self.assertEqual(elements, [])

    def test_sort_one_element_list(self):
        # Test sorting a list with one element
        elements = [5]
        bubble_sort_in_place(elements)
        self.assertEqual(elements, [5])

    def test_sort_sorted_list(self):
        # Test sorting a list that is already sorted
        elements = [1, 2, 3, 4, 5]
        bubble_sort_in_place(elements)
        self.assertEqual(elements, [1, 2, 3, 4, 5])

    def test_sort_reverse_sorted_list(self):
        # Test sorting a list that is sorted in reverse order
        elements = [5, 4, 3, 2, 1]
        bubble_sort_in_place(elements)
        self.assertEqual(elements, [1, 2, 3, 4, 5])

    def test_sort_unsorted_list(self):
        # Test sorting a list that is unsorted
        elements = [3, 5, 1, 4, 2]
        bubble_sort_in_place(elements)
        self.assertEqual(elements, [1, 2, 3, 4, 5])

    def test_sort_list_with_duplicates(self):
        # Test sorting a list with duplicate elements
        elements = [3, 5, 1, 4, 2, 3]
        bubble_sort_in_place(elements)
        self.assertEqual(elements, [1, 2, 3, 3, 4, 5])

    def test_raises_type_error(self):
        # Test that a TypeError is raised if the input is not a list
        with self.assertRaises(TypeError):
            bubble_sort_in_place("not a list")


if __name__ == '__main__':
    unittest.main()

File: kotlin/src/test/kotlin/sort/SortTest.kt
package sort

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

abstract class SortTest {

    abstract fun sort(elements: MutableList<Int>)

    @Test
    fun testSortEmptyList() {
        // Test sorting an empty list
        val elements = mutableListOf<Int>()
        sort(elements)
        Assertions.assertEquals(emptyList<Int>(), elements)
    }

    @Test
    fun testSortOneElementList() {
        // Test sorting a list with one element
        val elements = mutableListOf(5)
        sort(elements)
        Assertions.assertEquals(listOf(5), elements)
    }

    @Test
    fun testSortSortedList() {
        // Test sorting a list that is already sorted
        val elements = mutableListOf(1, 2, 3, 4, 5)
        sort(elements)
        Assertions.assertEquals(listOf(1, 2, 3, 4, 5), elements)
    }

    @Test
    fun testSortReverseSortedList() {
        // Test sorting a list that is sorted in reverse order
        val elements = mutableListOf(5, 4, 3, 2, 1)
        sort(elements)
        Assertions.assertEquals(listOf(1, 2, 3, 4, 5), elements)
    }

    @Test
    fun testSortUnsortedList() {
        // Test sorting a list that is unsorted
        val elements = mutableListOf(3, 5, 1, 4, 2)
        sort(elements)
        Assertions.assertEquals(listOf(1, 2, 3, 4, 5), elements)
    }

    @Test
    fun testSortListWithDuplicates() {
        // Test sorting a list with duplicate elements
        val elements = mutableListOf(3, 5, 1, 4, 2, 3)
        sort(elements)
        Assertions.assertEquals(listOf(1, 2, 3, 3, 4, 5), elements)
    }
}